fix: Report Service

DerekCrannaford opened the pull request:
Good UserWhere From AllCalls

Report UserWhere being called by array, string and 'empty'
Then breaking anyway due to incorrect quote escapes
Fixes for versatile calls & forced SQL escaping
Tested Support:
/report/index?module=
/report/reportAjaxCategorytoType?id=_
/report/reportAjaxListModules

refs: clients assert reports dont work

strongishllama reviewed (changes requested):
Looks good, just remove that commented var_dump on line 188. I'd be keen for @adam-buckley to approve this as well though before merging it. Just due to the importance of reports.

strongishllama reviewed (changes requested):
Looks good! Just add the comments and type hinting as we discussed, cheers.

DerekCrannaford commented:
Added commit to include patches on RunReport vs ExeReport for membership rules to match those of getUserWhere
Added SQLI unit test for getUserWhere : php cmfive.php tests unit report

adam-buckley commented on system/modules/report/models/ReportService.php:
```diff
@@ -27,7 +27,16 @@ public function getReportMembers($id)
+
+        foreach ($groups ?? [] as $group) {
+            if (AuthService::getInstance($this->w)->getUser($uid)->inGroup($group)) {
+                $conferred[] = $this->getObject("ReportMember", ["report_id" => $id, "user_id" => $group->id, "is_deleted" => 0]);
```
Thinking about this again, there's still a chance that the member return here is null. before the return below we should be calling array_filter over this conferred variable which will strip out null values

DerekCrannaford replied:
+1

adam-buckley commented on system/modules/report/models/ReportService.php:
```diff
@@ -144,25 +153,27 @@ public function getReportsbyUserWhere($id, $where)
         // need to get reports for me and my groups
         // me
-        $myid = [$id];
+        $myid[] = $this->_db->quote($id);
```
This is not the correct way to instantiate an array, even though this works it should be ```$myid = [$this->_db->quote($id)];```

DerekCrannaford replied:
OK

adam-buckley commented on system/modules/report/models/ReportService.php:
```diff
@@ -176,11 +187,45 @@ public function getReportsbyUserWhere($id, $where)
         return $this->fillObjects("Report", $rows);
     }
 
+     public function unitaryWhereToAndClause($where) {
```
Not PSR2 compliant

DerekCrannaford replied:
+1

DerekCrannaford commented:
Any suggested extensions/additions to unit test?
